### PR TITLE
Category permissions

### PR DESCRIPTION
**Pull request by rluck0419**
ITS FINALLY FINISHED!!!

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -2,4 +2,24 @@ class ApplicationController < ActionController::Base
   include Clearance::Controller
   protect_from_forgery with: :exception
 
+  def has_permission?(obj)
+    if obj.respond_to?(:user_id)
+      obj.user_id == current_user.id || current_user.admin? || obj.users_with_access.include?(current_user)
+    else
+      current_user.admin? || obj.users_with_access.include?(current_user)
+    end
```
Comment (jah2488) [inline app/controllers/application_controller.rb]: I'm confused why this is on both branches of the if statement:  `current_user.admin? || obj.users_with_access.include?(current_user)`

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -2,4 +2,24 @@ class ApplicationController < ActionController::Base
...
+
+  def current_permission?(obj)
+    parents = [obj.category.ancestors, obj.category].flatten
+    parents.each do |parent|
+      if has_permission?(parent)
+        return true
+      end
+    end
+
+    if has_permission?(photo)
```
Comment (jah2488) [inline app/controllers/application_controller.rb]: This will break? Where does `photo` come for?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -2,4 +2,24 @@ class ApplicationController < ActionController::Base
...
+    end
+  end
+
+  def current_permission?(obj)
+    parents = [obj.category.ancestors, obj.category].flatten
+    parents.each do |parent|
+      if has_permission?(parent)
+        return true
+      end
+    end
```
Comment (jah2488) [inline app/controllers/application_controller.rb]: This block of code could be refactored:

```
return true if (obj.category.ancestors << obj.category).any? { |cat| has_permission?(cat) }
```

File: app/controllers/notes_controller.rb [hunk 4]
```diff
@@ -5,20 +5,26 @@ def index
     else
       notes = Note.all
     end
-    render locals: { notes: notes }
+    perm_notes = []
+    notes.each do |note|
+      if current_permission?(note)
+        perm_notes << note
+      end
+    end
```
Comment (jah2488) [inline app/controllers/notes_controller.rb]: Remember, whenever you are filling an array through an each, you could refactor that to use `.map`

```
notes.map do |note|
  note if current_permission?(note)
end
```

File: app/controllers/photos_controller.rb [hunk 5]
```diff
@@ -5,20 +5,26 @@ def index
     else
       photos = Photo.all
     end
-    render locals: { photos: photos }
+    perm_photos = []
+    photos.each do |photo|
+      if current_permission?(photo)
+        perm_photos << photo
+      end
+    end
```
Comment (jah2488) [inline app/controllers/photos_controller.rb]: ditto above comment.

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -1,6 +1,18 @@
 module ApplicationHelper
+  def authorized_any?(array)
+    array.each do |item|
+      return true if authorized_show?(item)
+    end
+    return true if array.empty?
+    false
+  end
```
Comment (jah2488) [inline app/helpers/application_helper.rb]: I feel this block could also be refactored:

```
return true if array.any? { |item| authorized_show?(item) }

array.empty?
```

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,6 +1,18 @@
...
+
   def authorized_show?(object)
-    object.user_id == current_user.id || current_user.admin? || object.users_with_access.include?(current_user)
+    if object.user_id
+      object.user_id == current_user.id || current_user.admin? || object.users_with_access.include?(current_user)
+    else
+      current_user.admin? || object.users_with_access.include?(current_user)
+    end
```
Comment (jah2488) [inline app/helpers/application_helper.rb]: ditto above comment.

Why so much duplication?

File: app/views/categories/show.html.erb [hunk 8]
```diff
@@ -28,28 +28,38 @@
 <% if [category.children, category.notes, category.photos].all?(&:empty?) %>
   <p> This folder is empty! </p>
 <% else %>
-  <% category.children.each do |child| %>
-    <div class="category-name">
-      <% if [child.children, child.notes, child.photos].all?(&:empty?) %>
-        <%= link_to child.name + " (No items)", child %>
-      <% else %>
-        <%= link_to child.name + " (#{item_count(child)})", child %>
+  <% if authorized_any?((category.notes+category.photos)) %>
```
Comment (jah2488) [inline app/views/categories/show.html.erb]: Why double `(())`` and you need to have spaces around your operands

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (jah2488) [inline app/controllers/application_controller.rb]:
```
def has_permission?(obj)
  [
    current_user.admin?,                                         # User is admin?
    obj.users_with_access.include?(current_user),                # User has access to object
    obj.respond_to?(:user_id) && obj.user_id == current_user.id  # User created object
  ].any?
end
```

**Comment by jah2488** [comment]
👍

**Comment by MattRice12** [comment]
👍

**Pull request merged**